### PR TITLE
Add support for client_info for session continuation and caching

### PR DESCRIPTION
**Pull request by kpanwar**
Add support for client_info for session continuation and caching, per #67

### CODE REVIEW

File: src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs [hunk 1]
```diff
@@ -27,40 +27,25 @@
...
 using System.Collections.Generic;
-using System.Globalization;
 using System.Linq;
 using System.Text;
 
 namespace Microsoft.Identity.Client.Internal.Cache
 {
     /// <summary>
-    /// <see cref="TokenCacheKey" /> can be used with Linq to access items from the TokenCache dictionary.
+    /// <see cref="AccessTokenCacheKey" /> can be used with Linq to access items from the TokenCache dictionary.
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs]: "_AccessTokenCacheKey can be used with Linq to access items from the TokenCache dictionary_": are any of the TokenCacheKey classes actually used as keys in a dictionary? The dictionaries in _TokenCacheAccessor_ are keyed on _string_.

If the classes aren't used as keys then the _Equals_ and _GetHashCode_ methods and the associated tests aren't required.

In fact, it looks like the only method that really matters is "ToString", and this could be replaced with a static method "CalculateKey".

File: src/Microsoft.Identity.Client/Platforms/net45/NavigateErrorStatus.cs [hunk 2]
```diff
@@ -104,7 +104,7 @@ internal class NavigateErrorStatus
             },
             {
                 (int)NavigateErrorStatusCode.HTTP_STATUS_NOT_FOUND,
-                "The server has not found anything matching the requested URI (Uniform Scope Identifier)."
+                "The server has not found anything matching the requested URI (Uniform ScopeSet Identifier)."
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Platforms/net45/NavigateErrorStatus.cs]: ?"Resource"?. Same in the line 146

File: src/Microsoft.Identity.Client/TokenCache.cs [hunk 3]
```diff
@@ -323,8 +320,12 @@ internal ICollection<User> GetUsers(string clientId)
                 IDictionary<string, User> allUsers = new Dictionary<string, User>();
                 foreach (RefreshTokenCacheItem item in tokenCacheItems)
                 {
-                    User user = new User(item.User);
-                    allUsers[item.HomeObjectId] = user;
+                    if (environment.Equals(
+                        item.Environment, StringComparison.OrdinalIgnoreCase))
```
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/TokenCache.cs]: It would be better to create a standard helper function _AreEnvironmentsEqual(env1, env2)_ that does the comparison. That avoids the risk of different _StringComparison_ kinds being used in different places.
The same applies to the comparison of _ClientId_.

File: tests/Test.MSAL.NET.Unit/CacheTests/RefreshTokenCacheKeyTests.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Test.MSAL.NET.Unit.CacheTests
+{
+    [TestClass]
+    public class RefreshTokenCacheKeyTests
+    {
+        [TestMethod]
+        [TestCategory("RefreshTokenCacheKeyTests")]
+        public void ConstructorTest()
```
Comment (duncanpMS) [inline tests/Test.MSAL.NET.Unit/CacheTests/RefreshTokenCacheKeyTests.cs]: Related to an earlier comment: unless _RefreshTokenCacheKey_ objects are directly compared against each other or used as keys in dictionary then there's no point in overriding _Equals_ and _GetHashCode_, so these tests wouldn't be needed either.

File: tests/Test.MSAL.NET.Unit/ClientInfoTests.cs [hunk 5]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
```
Comment (duncanpMS) [inline tests/Test.MSAL.NET.Unit/ClientInfoTests.cs]: Copyright

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs [hunk 6]
```diff
@@ -0,0 +1,73 @@
+﻿using System.Text;
+
+namespace Microsoft.Identity.Client.Internal.Cache
+{
+    class RefreshTokenCacheKey : TokenCacheKeyBase
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs]: scopes are not stored for refresh tokens.

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs [hunk 7]
```diff
@@ -0,0 +1,73 @@
+﻿using System.Text;
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs]: fixed.

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        public string Environment { get; }
+
+        /// <summary>
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.Append(MsalHelpers.EncodeToBase64Url(Environment) + "$");
+            stringBuilder.Append(MsalHelpers.EncodeToBase64Url(ClientId) + "$");
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs]: fixed

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+                    && UserIdentifier == other.UserIdentifier);
+        }
+
+        /// <summary>
+        /// Returns the hash code for this AccessTokenCacheKey.
+        /// </summary>
+        /// <returns>
+        /// A 32-bit signed integer hash code.
+        /// </returns>
+        public override int GetHashCode()
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs]: fixed

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs [hunk 10]
```diff
@@ -33,21 +34,80 @@ namespace Microsoft.Identity.Client.Internal.Cache
...
         }
 
+        [DataMember(Name = "environment")]
+        public string Environment { get; set; }
+
+        [DataMember(Name = "uid")]
+        public string Uid { get; set; }
+
+        [DataMember(Name = "utid")]
+        public string Utid { get; set; }
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs]: cacheitem classes are internal. they might get rolled over from STS

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs [hunk 11]
```diff
@@ -33,21 +34,80 @@ namespace Microsoft.Identity.Client.Internal.Cache
...
+                MsalHelpers.EncodeToBase64Url(Utid));
+        }
+
+        // This method is called after the object 
+        // is completely deserialized.
+        [OnDeserialized]
+        void OnDeserialized(StreamingContext context)
         {
-            return new TokenCacheKey(null, null, ClientId, User.HomeObjectId);
+            User = User.Create(DisplayableId, Name, IdentityProvider,
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs]: We can create User object from artifacts stored in the cache. Goal is to preserve things that are returned from the server. User is simply a representation in a particular fashion of the data stored.

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs [hunk 12]
```diff
@@ -36,78 +35,23 @@ namespace Microsoft.Identity.Client.Internal.Cache
...
         {
         }
 
-        internal IdToken IdToken { get; set; }
-
-        /// <summary>
-        /// Gets the ClientId.
-        /// </summary>
         [DataMember(Name = "client_id")]
         public string ClientId { get; set; }
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: I can address this in future iterations. I do plan on doing an OOP sanity check across the board.

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs [hunk 13]
```diff
@@ -36,78 +35,23 @@ namespace Microsoft.Identity.Client.Internal.Cache
...
-                UniqueId = IdToken.ObjectId;
-                if (UniqueId == null)
-                {
-                    UniqueId = IdToken.Subject;
-                }
-
-                User = User.CreateFromIdToken(IdToken);
-            }
-        }
+        public abstract string GetUserIdentifier();
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: I need to call it function in other classes.

File: src/Microsoft.Identity.Client/Internal/Cache/TokenCacheKeyBase.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            ClientId = clientId;
+            UserIdentifier = userIdentifier;
+        }
+
+        public string ClientId { get; set; }
+
+        public string UserIdentifier { get; set; }
+
+        public bool Equals(string string1, string string2)
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/TokenCacheKeyBase.cs]: it can be used for more comparisons later. I made it protected.

File: src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheItem.cs [hunk 15]
```diff
@@ -60,54 +120,49 @@ public DateTimeOffset ExpiresOn
...
         {
-            return new TokenCacheKey(Authority, Scope, ClientId, User);
+            ScopeSet = Scope.AsSet();
+            IdToken = IdToken.Parse(RawIdToken);
+            ClientInfo = ClientInfo.Parse(RawClientInfo);
+            if (IdToken != null)
+            {
+                User = User.Create(IdToken.PreferredUsername, IdToken.Name, IdToken.Issuer,
+                    GetUserIdentifier());
+            }
```
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheItem.cs]: added CreateDerivedProperties() as a separate function

File: src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs (hunk 1 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs]: You are right, we do need only ToString(). I would prefer to call it tostring because it is the string representation of the object and more natural to use.

File: src/Microsoft.Identity.Client/Platforms/net45/NavigateErrorStatus.cs (hunk 2 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Platforms/net45/NavigateErrorStatus.cs]: it should be scope. We don't have resources anymore. but this file a victim of rename. I will revert it

File: tests/Test.MSAL.NET.Unit/ClientInfoTests.cs (hunk 5 above)
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit/ClientInfoTests.cs]: fixed

File: tests/Test.MSAL.NET.Unit/CacheTests/RefreshTokenCacheKeyTests.cs (hunk 4 above)
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit/CacheTests/RefreshTokenCacheKeyTests.cs]: fixed.

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs [hunk 16]
```diff
@@ -115,7 +115,7 @@ public IEnumerable<User> Users
                     return new List<User>();
                 }
 
-                return UserTokenCache.GetUsers(ClientId);
+                return UserTokenCache.GetUsers(new Uri(Authority).Host);
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: this is a change that is gonna affect all the platforms. I'm not seeing it documented anywhere that we're going to retrieve users with authority host instead of client id, first question is why we are doing it? Second is an ask, can you file an issue on all the platforms so we can make sure that all of us make the change.

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs [hunk 17]
```diff
@@ -36,78 +36,32 @@ namespace Microsoft.Identity.Client.Internal.Cache
     [DataContract]
     internal abstract class BaseTokenCacheItem
     {
-        protected RequestContext RequestContext;
-
         /// <summary>
         /// Default constructor.
         /// </summary>
-        internal BaseTokenCacheItem(string authority, string clientId, TokenResponse response)
+        public BaseTokenCacheItem(string clientId)
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: I noticed that you changed the visibility of a whole bunch of the internal class constructors, technically it doesn't matter for the visibility of internal class. Just out of curiosity of the reason of changing it.

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 16 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: Client id is still being used. There was no need to pass it in the function as it was present as a class property. We just added an additional constraint.

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs [hunk 18]
```diff
@@ -36,78 +36,32 @@ namespace Microsoft.Identity.Client.Internal.Cache
...
-                IdToken = IdToken.Parse(RawIdToken, RequestContext);
-                UniqueId = IdToken.ObjectId;
-                if (UniqueId == null)
-                {
-                    UniqueId = IdToken.Subject;
-                }
-
-                User = User.CreateFromIdToken(IdToken);
-            }
+            return string.Format(CultureInfo.InvariantCulture, "{0}.{1}", MsalHelpers.EncodeToBase64Url(ClientInfo?.UniqueIdentifier),
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: if I read the correctly, client_info is not a required field for TokenResponse. We'll run into null pointer exception if client_info is not returned. Also given that uid and utid are the key part for us to construct key, we should probably consider failing the request sent to token endpoint.

File: src/Microsoft.Identity.Client/Internal/ClientInfo.cs [hunk 19]
```diff
@@ -42,16 +43,16 @@ internal class ClientInfo
...
 
         [DataMember(Name = ClientInfoClaim.UnqiueTenantIdentifier, IsRequired = false)]
-        public string UnqiueTenantIdentifier { get; set; }
+        public string UniqueTenantIdentifier { get; set; }
 
         public static ClientInfo Parse(string clientInfo)
         {
             if (string.IsNullOrEmpty(clientInfo))
             {
                 return null;
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/ClientInfo.cs]: :) what are we going to use for uid and utid if it's null?

**Comment by kpanwar** [comment]
@duncanpMS - I will add more detailed comments next time.

File: src/Microsoft.Identity.Client/Internal/IdToken.cs [hunk 20]
```diff
@@ -90,20 +90,19 @@ public static IdToken Parse(string idToken, RequestContext requestContext)
                             idTokenBody = (IdToken) serializer.ReadObject(stream);
                         }
                     }
-                    catch (SerializationException ex)
+                    catch (SerializationException)
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/IdToken.cs]: you only need to exception class here? Also why the logging is removed?

File: src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs [hunk 21]
```diff
@@ -127,9 +127,10 @@ private Uri CreateAuthorizationUri(bool addVerifier = false, bool addState = fal
             if (addVerifier)
             {
                 _codeVerifier = PlatformPlugin.CryptographyHelper.GenerateCodeVerifier();
-                string codeVerifierHash = PlatformPlugin.CryptographyHelper.CreateSha256Hash(_codeVerifier);
+                string codeVerifierHash = MsalHelpers.EncodeToBase64Url(
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: I don't understand why the change is made. Actually i prefer the code before the change. Generate the code verifier hash first, and then encoded it when we set the request params, this looks more reasonable to me, and help on debugging.

File: src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs [hunk 22]
```diff
@@ -81,14 +81,7 @@ protected override void SetAdditionalRequestParameters(OAuth2Client client)
 
                 RequestContext.Logger.Verbose("Refreshing access token...");
                 await base.SendTokenRequestAsync().ConfigureAwait(false);
-
-                if (Response.IdToken == null)
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: Wooohoo! does it mean that we have an agreement with service team that id token will always be returned?

File: src/Microsoft.Identity.Client/TokenCache.cs [hunk 23]
```diff
@@ -260,7 +260,9 @@ internal RefreshTokenCacheItem FindRefreshToken(AuthenticationRequestParameters
         {
             lock (LockObject)
             {
-                TokenCacheKey key = new TokenCacheKey(null, null, requestParam.ClientId, requestParam.User?.HomeObjectId);
+                RefreshTokenCacheKey key = new RefreshTokenCacheKey(
+                    new Uri(requestParam.Authority.CanonicalAuthority).Host, requestParam.ClientId,
```
Comment (weijjia) [inline src/Microsoft.Identity.Client/TokenCache.cs]: can't the authority class expose a method i.e. getHost instead of the new Uri used in multiple places?

File: src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs (hunk 21 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Requests/InteractiveRequest.cs]: this is because of [https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/289](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/289)
With the next fix, this code will look exactly how it looked before

---
In reply to: [109345485](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109345485) [](ancestors = 109345485)

File: src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs (hunk 22 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Requests/SilentRequest.cs]: yes. thats the whole point of openid scope.

---
In reply to: [109345584](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109345584) [](ancestors = 109345584)

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs (hunk 17 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: for serialization/deserialization

---
In reply to: [109344344](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109344344) [](ancestors = 109344344)

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs (hunk 18 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: We don't know if ADFS will be supported. They have a lot of features to add for us. We should not add fallbacks for STS which is not going to be supported in the near/distant future.

---
In reply to: [109344068](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109344068) [](ancestors = 109344068)

Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: This does not throw null pointer. It results in a user id being "." and equates to single user at a time STS (like ADFS).

---
In reply to: [109344983](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109344983) [](ancestors = 109344983)

File: src/Microsoft.Identity.Client/ClientApplicationBase.cs (hunk 16 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/ClientApplicationBase.cs]: see the bug filed in your repo.

---
In reply to: [109343735](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109343735) [](ancestors = 109343735)

File: src/Microsoft.Identity.Client/Internal/IdToken.cs (hunk 20 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/IdToken.cs]: I will remove the exception. it should be bubbled up.

---
In reply to: [109345288](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109345288) [](ancestors = 109345288)

File: src/Microsoft.Identity.Client/TokenCache.cs (hunk 23 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/TokenCache.cs]: fixed

---
In reply to: [109345765](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/282#discussion_r109345765) [](ancestors = 109345765)

File: tests/Test.MSAL.NET.Unit.PublicApi/ReadMe.txt [hunk 24]
```diff
@@ -1,7 +0,0 @@
-﻿Unit tests for the public API only
-----------------------------------
-
-It is important this project is *not* able to see the internals of the product assemblies,
-which is why these tests are in a separate project (the product internals are visible to
-"Test.MSAL.Net.Unit").
```
Comment (duncanpMS) [inline tests/Test.MSAL.NET.Unit.PublicApi/ReadMe.txt]: What was the reason for merging the two test projects? This one was deliberately separated to make sure it could only be testing the public API.

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs (hunk 6 above)
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheKey.cs]: I meant C# scope keyword - _public_, _internal_ etc.

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs (hunk 10 above)
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs]: How is that going affect caching? If the properties are changed after the item has been put into the cache then the next time the cache key is calculated it will give a different value -> the existing cache entry won't be found.

File: tests/Test.MSAL.NET.Unit.PublicApi/ReadMe.txt (hunk 24 above)
Comment (kpanwar) [inline tests/Test.MSAL.NET.Unit.PublicApi/ReadMe.txt]: Public API can also be tested in a single project. It is an additional project overhead that I's rather not maintain. You can create separate test files if needed (i don't think its necessary).

File: src/Microsoft.Identity.Client/Internal/Cache/TokenCacheKeyBase.cs (hunk 14 above)
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Internal/Cache/TokenCacheKeyBase.cs]: Method and parameter names are important. _Equals_ gives no clue about the what it should be used to compare - client ids, authorities, environments? Put yourself in the position of someone else maintaining the code in three years time: how are they going to know what this method is for?

File: src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs (hunk 10 above)
Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs]: Yep, this is exactly the behavior that will happen. It is a an edge case that should never happen unless there is a security issue and MSA needs to roll their salt/algorithm. This only applies to MSA btw.

Comment (kpanwar) [inline src/Microsoft.Identity.Client/Internal/Cache/RefreshTokenCacheItem.cs]: Yep, this is exactly the behavior that will happen. It is a an edge case that should never happen unless there is a security issue and MSA needs to roll their salt/algorithm. This only applies to MSA btw.

File: src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs (hunk 1 above)
Comment (duncanpMS) [inline src/Microsoft.Identity.Client/Internal/Cache/AccessTokenCacheKey.cs]: A different method name would make it clearer that the string returned has a special property - that it provides a logical identifier that can be used for caching and comparisons.
That isn't normally the case for overrides of _ToString_. It's overridden for a variety of different reasons e.g. to provide a user-friendly description, debug info, or a serialized form of the class. Anyone looking at the code will not immediately expect _ToString_ to return a key value. You could add a doc comment to explain it, but they can get out of date and often people don't read them. Using a different method name can make it immediately obvious without requiring any further explanation.

File: src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs (hunk 18 above)
Comment (weijjia) [inline src/Microsoft.Identity.Client/Internal/Cache/BaseTokenCacheItem.cs]: ok, i should add a test case for the scenario when client info is not returned.
